include/dice/template-library/polymorphic_allocator.hpp

```diff
@@ -197,6 +197,583 @@ namespace dice::template_library {
+			std::swap(lhs.alloc_, rhs.alloc_);
+		}
+
+		constexpr polymorphic_allocator select_on_container_copy_construction() const {
+			return polymorphic_allocator{std::allocator_traits<Allocator<T>>::select_on_container_copy_construction(alloc_)};
+		}
+
+		/**
+		 * Checks if *this currently holds an instance of UAllocator
+		 */
+		template<typename UAllocator>
+		[[nodiscard]] constexpr bool holds_allocator() const noexcept {
+			return std::is_same_v<UAllocator, Allocator<T>>;
+		}
+
+		/**
+		 * Checks if *this currently holds an instance of UAllocator<T>
+		 */
+		template<template<typename> typename UAllocator>
+		[[nodiscard]] constexpr bool holds_allocator() const noexcept {
+			return std::is_same_v<UAllocator<T>, Allocator<T>>;
+		}
+	};
+
+	namespace discriminant2_detail {
+		enum struct Discriminant : bool {
+			First,
+			Second,
+		};
+	} // namespace discriminant2_detail
+
+	template<typename T, template<typename> typename Allocator1, template<typename> typename Allocator2>
+	struct polymorphic_allocator<T, Allocator1, Allocator2> {
+		static_assert(!std::is_same_v<Allocator1<T>, Allocator2<T>>,
+					  "Allocator types must only occur exactly once in the parameter list");
+
+		using value_type = typename detail_pmr::same_type<typename std::allocator_traits<Allocator1<T>>::value_type,
+														  typename std::allocator_traits<Allocator2<T>>::value_type>::type;
+
+		using pointer = typename detail_pmr::same_type<typename std::allocator_traits<Allocator1<T>>::pointer,
+													   typename std::allocator_traits<Allocator2<T>>::pointer>::type;
+
+		using const_pointer = typename detail_pmr::same_type<typename std::allocator_traits<Allocator1<T>>::const_pointer,
+															 typename std::allocator_traits<Allocator2<T>>::const_pointer>::type;
+
+		using void_pointer = typename detail_pmr::same_type<typename std::allocator_traits<Allocator1<T>>::void_pointer,
+															typename std::allocator_traits<Allocator2<T>>::void_pointer>::type;
+
+		using const_void_pointer = typename detail_pmr::same_type<typename std::allocator_traits<Allocator1<T>>::const_void_pointer,
+																  typename std::allocator_traits<Allocator2<T>>::const_void_pointer>::type;
+
+		using propagate_on_container_copy_assignment = std::bool_constant<(std::allocator_traits<Allocator1<T>>::propagate_on_container_copy_assignment::value
+																			|| std::allocator_traits<Allocator2<T>>::propagate_on_container_copy_assignment::value)>;
+
+		using propagate_on_container_move_assignment = std::bool_constant<(std::allocator_traits<Allocator1<T>>::propagate_on_container_move_assignment::value
+																			|| std::allocator_traits<Allocator2<T>>::propagate_on_container_move_assignment::value)>;
+
+		using propagate_on_container_swap = std::bool_constant<(std::allocator_traits<Allocator1<T>>::propagate_on_container_swap::value
+																 || std::allocator_traits<Allocator2<T>>::propagate_on_container_swap::value)>;
```

Comment:
Thats a good question. 
I guess the more fundamental question is when does an allocator not want to propagate?

Reply:
> And why OR instead of AND between the allocators. Wouldn't AND be the more safe and conservative option?

I think I got that idea from https://en.cppreference.com/w/cpp/memory/scoped_allocator_adaptor because it uses or.

Reply:
> Thats a good question. I guess the more fundamental question is: when does an allocator not want to propagate?

I can think of situations with allocators for VRAM on different GPUs. But nothing that is relevant for what we do ATM. :)

> I think I got that idea from https://en.cppreference.com/w/cpp/memory/scoped_allocator_adaptor because it uses or.

Then let's stick to that. Maybe add a doctstring (to the general case) that we followed that.

Reply:
added comment